destination/writer/writer.go

```diff
@@ -181,264 +209,355 @@ func (s *SnowflakeCSV) Write(ctx context.Context, records []sdk.Record) (int, er
-		if len(schema) != len(snowflakeSchema) {
-			sdk.Logger(ctx).Debug().Msgf("Snowflake Table Schema (%+v)", snowflakeSchema)
-			sdk.Logger(ctx).Debug().Msgf("Source Table Schema (%+v)", schema)
-
-			return errors.Errorf("table already exists on snowflake, source schema number of "+
-				"columns %d doesn't match destination table %d",
-				len(schema),
-				len(snowflakeSchema))
+
+		for i, col := range t.Columns {
+			if col.Name == colName {
+				t.PrimaryKeys = append(t.PrimaryKeys, t.Columns[i])
+				continue
+			}
+			return snowflake.Table{}, fmt.Errorf("failed to find primary key column %q in table %q", colName, t.Name)
 		}
+	}
+	if selectPrimaryKeysResult.Err() != nil {
+		return snowflake.Table{}, fmt.Errorf("failed to iterate over primary keys: %w", selectPrimaryKeysResult.Err())
+	}
+
+	if err := tx.Commit(); err != nil {
+		return snowflake.Table{}, fmt.Errorf("failed to commit transaction: %w", err)
+	}
+
+	return t, nil
+}
+
+func (s *SnowflakeCSV) CheckTable(
+	ctx context.Context,
+	operation sdk.Operation,
+	snowflakeSchema snowflake.Table,
+	recordSchema snowflake.Table,
+) error {
+	// if snowflake schema is empty, no need to check anything
+	if len(snowflakeSchema.Columns) == 0 {
+		return ErrTableNotFound
+	}
+
+	sdk.Logger(ctx).Debug().Msgf("Existing Table Schema (%+v)", snowflakeSchema.Columns)
+	sdk.Logger(ctx).Debug().Msgf("Connector Generated Schema (%+v)", recordSchema.Columns)
 
-		err := schemaMatches(schema, snowflakeSchema)
+	// Primary keys need to match either way.
+	err := columnsMatch(recordSchema.PrimaryKeys, snowflakeSchema.PrimaryKeys)
+	if err != nil {
+		return fmt.Errorf("primary keys do not match: %w", err)
+	}
+	if operation != sdk.OperationDelete {
+		// if operation is not delete, all columns should match
+		err := columnsMatch(recordSchema.Columns, snowflakeSchema.Columns)
 		if err != nil {
-			return err
+			return fmt.Errorf("columns do not match: %w", err)
 		}
 	}
 
 	return nil
 }
 
-func schemaMatches(schema, snowflakeSchema map[string]string) error {
-	for k, v := range schema {
-		lowerK := strings.ToLower(k)
-		v2, ok := snowflakeSchema[lowerK]
-		if !ok {
-			return errors.Errorf("table already exists on snowflake, column %s doesn't exist on destination table ", k)
-		}
-		if !strings.EqualFold(v, v2) {
-			return errors.Errorf("table already exists on snowflake, column %s with datatype %s "+
-				"doesn't match on destination table of datatype %s ", k, v, v2)
+func columnsMatch(cols1, cols2 []snowflake.Column) error {
+	if len(cols1) != len(cols2) {
+		return fmt.Errorf("number of columns doesn't match (%d != %d)", len(cols1), len(cols2))
+	}
+	for k, col1 := range cols1 {
+		col2 := cols2[k]
+		if !strings.EqualFold(col1.Name, col2.Name) {
+			return fmt.Errorf("column %d doesn't match (%s:%T != %s:%T)", k, col1.Name, col1.DataType, col2.Name, col2.DataType)
 		}
+		// TODO check data type? what if the source record has nil values and we don't know types?
```

Comment:
This is a bit shady - we compare the schema (columns) we extracted from the first record with the snowflake table schema. The problem is if some fields are nullable and/or the record is partially populated, because we expect the exact same number of columns and the same order of columns.

I'm pretty sure the ordering is not guaranteed right now, as a schema fetched from the record will have connector columns in front (`operation`, `created_at` etc.), while the columns in a snowflake table schema will be simply ordered alphabetically. Some work is needed to make sure we have consistent ordering.